### PR TITLE
Always inform the failure detector when responses come in from the network client

### PR DESCRIPTION
**Pull request by pnarayanan**
When NetworkClient hands over a response, the operation managers may
ignore it - if the request timed out internally, or if the operation
itself completed. However, the response handler responsible for failure
detection should be notified about these responses, regardless.

--
*Coverage*
```
DeleteManager           100% (2/ 2)     100% (11/ 11)   93% (80/ 86)
DeleteOperation         100% (3/ 3)     100% (21/ 21)   95.4% (103/ 108)
GetBlobInfoOperation    100% (2/ 2)     100% (14/ 14)   92.7% (101/ 109)
GetBlobOperation        100% (7/ 7)     89.8% (44/ 49)  90.4% (227/ 251)
GetManager              100% (2/ 2)     100% (13/ 13)   78.6% (66/ 84)
GetOperation            100% (1/ 1)     100% (10/ 10)   100% (24/ 24)
NonBlockingRouter       100% (3/ 3)     100% (28/ 28)   86.8% (138/ 159)
PutManager              100% (4/ 4)     100% (17/ 17)   89.9% (133/ 148)
PutOperation            100% (7/ 7)     100% (62/ 62)   92.4% (292/ 316)
RouterRequestInfo       100% (1/ 1)     100% (2/ 2)     100% (3/ 3)
```

### CODE REVIEW

File: ambry-router/src/main/java/com.github.ambry.router/DeleteManager.java [hunk 1]
```diff
@@ -169,6 +177,30 @@ void handleResponse(ResponseInfo responseInfo) {
...
+    NetworkClientErrorCode networkClientErrorCode = responseInfo.getError();
+    if (networkClientErrorCode != null) {
+      responseHandler.onRequestResponseException(replicaId, new IOException("NetworkClient error"));
+    } else {
+      try {
+        deleteResponse =
+            DeleteResponse.readFrom(new DataInputStream(new ByteBufferInputStream(responseInfo.getResponse())));
+        responseHandler.onRequestResponseError(replicaId, deleteResponse.getError());
+      } catch (Exception e) {
+        // Ignore. There is no value in notifying the response handler.
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/DeleteManager.java]: metric?

File: ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java [hunk 2]
```diff
@@ -642,7 +646,6 @@ private void processGetBlobResponse(GetRequestInfo getRequestInfo, GetResponse g
               "received: " + partitionsInResponse, RouterErrorCode.UnexpectedInternalError);
         } else {
           getError = getResponse.getPartitionResponseInfoList().get(0).getErrorCode();
-          responseHandler.onRequestResponseError(getRequestInfo.replicaId, getError);
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: why is this removed?

Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java]: Never mind - saw the handling code does this already.

File: ambry-router/src/main/java/com.github.ambry.router/GetManager.java [hunk 3]
```diff
@@ -203,6 +212,34 @@ void handleResponse(ResponseInfo responseInfo) {
...
+      try {
+        getResponse = GetResponse
+            .readFrom(new DataInputStream(new ByteBufferInputStream(responseInfo.getResponse())), clusterMap);
+        ServerErrorCode serverError = getResponse.getError();
+        if (serverError == ServerErrorCode.No_Error) {
+          serverError = getResponse.getPartitionResponseInfoList().get(0).getErrorCode();
+        }
+        responseHandler.onRequestResponseError(replicaId, serverError);
+      } catch (Exception e) {
+        // Ignore. There is no value in notifying the response handler.
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/GetManager.java]: metric?

File: ambry-router/src/main/java/com.github.ambry.router/PutManager.java [hunk 4]
```diff
@@ -205,6 +213,29 @@ void handleResponse(ResponseInfo responseInfo) {
...
+    ReplicaId replicaId = ((RouterRequestInfo) responseInfo.getRequestInfo()).getReplicaId();
+    NetworkClientErrorCode networkClientErrorCode = responseInfo.getError();
+    if (networkClientErrorCode != null) {
+      responseHandler.onRequestResponseException(replicaId, new IOException("NetworkClient error"));
+    } else {
+      try {
+        putResponse = PutResponse.readFrom(new DataInputStream(new ByteBufferInputStream(responseInfo.getResponse())));
+        responseHandler.onRequestResponseError(replicaId, putResponse.getError());
+      } catch (Exception e) {
+        // Ignore. There is no value in notifying the response handler.
```
Comment (vgkholla) [inline ambry-router/src/main/java/com.github.ambry.router/PutManager.java]: metric?

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java [hunk 5]
```diff
@@ -157,6 +182,277 @@ public void testMultipleScalingUnit()
...
+          failedReplicaIds.get(0));
+      Assert.assertEquals("Successful notification should have arrived for replicas that were up",
+          DELETE_REQUEST_PARALLELISM - 1, successfulResponseCount.get());
+      Assert.assertFalse("There should be no notifications of any other kind", invalidResponse.get());
+    }
+    failedReplicaIds.clear();
+    successfulResponseCount.set(0);
+    invalidResponse.set(false);
+    mockSelectorState.set(MockSelectorState.Good);
+  }
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: there is quite a bit of repeated code in these three functions - would recommend trying to make it leaner and reuse code.

**Comment by vgkholla** [comment]
LGTM. Minor comments. 

Could you post unit test coverage? And confirm that all the new lines are covered?

Also, is there any way we can test this in a local deployment before committing?

File: ambry-router/src/main/java/com.github.ambry.router/DeleteManager.java (hunk 1 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/DeleteManager.java]: added.

File: ambry-router/src/main/java/com.github.ambry.router/GetManager.java (hunk 3 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/GetManager.java]: added.

File: ambry-router/src/main/java/com.github.ambry.router/PutManager.java (hunk 4 above)
Comment (pnarayanan) [inline ambry-router/src/main/java/com.github.ambry.router/PutManager.java]: added.

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java (hunk 5 above)
Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: done.

**Comment by pnarayanan** [comment]
Had tested and verified this in a test cluster before putting this up.
Couldn't put up coverage due to editor weirdness. Will try again. Should be close to 100%.

**Comment by pnarayanan** [comment]
@vgkholla addressed comments, please take a look.

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java [hunk 6]
```diff
@@ -275,145 +283,29 @@ public void onRequestResponseError(ReplicaId replicaId, ServerErrorCode serverEr
 
   /**
    * Test that the failure handler is notified for put responses.
-   * @param putManager the {@link PutManager}
+   * @param opHelper the {@link OperationType}
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: minor: the {@link OperationHelper}

**Comment by vgkholla** [comment]
Nicely done. LGTM

Please confirm build, formatting and 100% unit test coverage of the new lines and we should be good to go.

**Comment by pnarayanan** [comment]
Updated coverage percentage. I could not verify unit test coverage for all code due to coverage report issues, but all newly added code should be getting covered except for certain exception paths. Builds on both OS X and Linux.

**Comment by pnarayanan** [comment]
@vgkholla Verified coverage for newly added lines and added an unrelated test for response deserialization error handling. Please take a look.

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java [hunk 7]
```diff
@@ -242,47 +242,51 @@ public void onRequestResponseError(ReplicaId replicaId, ServerErrorCode serverEr
...
     putManager.close();
     getManager.close();
     deleteManager.close();
   }
 
   /**
-   * Test that the failure handler is notified for put responses.
+   * Test that failure detector is correctly notified for all responses regardless of the order in which successful
+   * and failed responses arrive.
    * @param opHelper the {@link OperationHelper}
```
Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: minor: doc params need changing.

**Comment by vgkholla** [comment]
Just one documentation comment. Please resolve it in your next PR. Merging this for now.

File: ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java (hunk 7 above)
Comment (pnarayanan) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: do they? which ones exactly?

Comment (vgkholla) [inline ambry-router/src/test/java/com.github.ambry.router/NonBlockingRouterTest.java]: I thought `invalidResponse` was added - my bad. Nothing to change then !
